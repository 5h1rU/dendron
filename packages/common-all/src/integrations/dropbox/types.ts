import { files } from "dropbox";

export type MetadataReference =
  | files.FileMetadataReference
  | files.FolderMetadataReference;

export interface ListFolderResultSimple {
  /**
   * The files and (direct) subfolders in the folder.
   */
  entries: Array<MetadataReference>;
  /**
   * Pass the cursor into listFolderContinue() to see what's changed in the
   * folder since your previous query.
   */
  cursor: files.ListFolderCursor;
  /**
   * If true, then there are more entries available. Pass the cursor to
   * listFolderContinue() to retrieve the rest.
   */
  has_more: boolean;
}
