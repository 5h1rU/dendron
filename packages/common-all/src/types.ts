// Global Types

import { Note, Schema } from "./node";

export type Stage = "dev" | "prod" | "test";

// === Node Types

// --- Primitives
export type NoteData = {
  schemaId?: string;
};
export type SchemaData = {
  namespace?: boolean;
};
export type DNodeData = SchemaData | NoteData;
export type IDNodeType = "note" | "schema";

export type QueryMode = IDNodeType;

// --- Nodes
export type DNodeRawOpts<T extends DNodeData> = {
  id?: string;
  title?: string;
  desc?: string;
  stub?: boolean;
  updated?: string;
  created?: string;
  /**
   * Same as uri
   */
  fname?: string;
  parent?: string | null | "root";
  children?: string[];
  body?: string;
  data?: T;
};
export type DNodeRawProps<T extends DNodeData = DNodeData> = Required<
  DNodeRawOpts<T>
>;

export type IDNodeOpts<T = DNodeData> = Omit<
  DNodeRawOpts<T>,
  "parent" | "children"
> & {
  type: IDNodeType;
  parent?: IDNode<T> | null;
  children?: IDNode<T>[];
};
export type IDNodeProps<T = DNodeData> = Required<IDNodeOpts<T>>;

export type IDNode<T = DNodeData> = IDNodeProps<T> & {
  // generated
  nodes: IDNode<T>[];
  // NOTE: currently equal to `fname`
  path: string;
  basename: string;
  logicalPath: string;
  // how to display path in lookup
  // used to special case `root` node to be empty string
  queryPath: string;
  // the root of domain
  domain: IDNode<T>;
  // absolute url to node id
  url: string;
  label: string;

  equal(node: IDNode<T>): boolean;
  // match(identifier: string): boolean;
  addChild(node: IDNode<T>): void;
  /**
   * Render body for rich-markdown-editor
   */
  renderBody(): string;
  toDocument(): any;
  toRawProps(): DNodeRawProps<T>;
  toRawPropsRecursive(): DNodeRawProps<T>[];
  validate(): boolean;
};
export type DNodeRawDict<T = DNodeData> = { [id: string]: DNodeRawProps<T> };
export type DNodeDict<T = DNodeData> = { [id: string]: IDNode<T> };

// --- Notes
export type NoteRawProps = DNodeRawProps<NoteData>;
export type INoteOpts = Omit<IDNodeOpts<NoteData>, "type">;
export type INoteProps = Required<INoteOpts>;
export type INote = INoteProps;
export type NoteDict = { [id: string]: Note };

// --- Schema
export type SchemaRawOpts = DNodeRawOpts<SchemaData> &
  Required<Pick<DNodeRawOpts<SchemaData>, "id" | "fname">>;
export type SchemaRawProps = DNodeRawProps<SchemaData>;
export type ISchemaOpts = Omit<IDNodeOpts<SchemaData>, "type">;
export type ISchemaProps = Required<ISchemaOpts>;
export type ISchema = ISchemaProps;
export type SchemaDict = { [id: string]: Schema };

// === Engine Types
export interface Resp<T> {
  data: T;
  error?: Error | null;
}

export type EngineGetResp<T = DNodeData> = Resp<IDNode<T>>;
export type EngineQueryResp<T = DNodeData> = Resp<IDNode<T>[]>;
export type StoreGetResp<T = DNodeData> = Resp<DNodeRawProps<T>>;
export type StoreQueryResp<T = DNodeData> = Resp<DNodeRawProps<T>[]>;

export interface Scope {
  username: string;
}

export interface NodeWriteOpts {
  /**
   * If newNode, will add it to the parent
   */
  newNode?: boolean;
  body?: string;
  /**
   * See QueryOpts.stub
   */
  stub?: boolean;
}

export interface QueryOpts {
  fullNode?: boolean;
  /**
   * Just get one result
   */
  queryOne?: boolean;
  /**
   * Use with `createIfNew`
   * If true, create a stub node
   */
  stub?: boolean;
  /**
   * If node does not exist, create it?
   */
  createIfNew?: boolean;
  // --- hints
  // DEPPRECATE
  webClient?: boolean;
  initialQuery?: boolean;
  mode?: QueryMode;
}

export interface DEngineParser<TOpts = any> {
  // parse: <T>(content: any, mode: QueryMode, opts: TOpts) => DNodeRawProps<T>[];
  parseSchema(data: any, opts: TOpts): SchemaRawProps[];
  parseNote(data: any, opts: TOpts): NoteRawProps[];
}

export type DEngineStoreWriteOpts = {
  /**
   * If set, don't write to file
   */
  stub?: boolean;
};

export interface DEngineStore<T = DNodeData, O = any> {
  opts: O;
  // fetchInitial: () => DNodeDict;
  delete: (id: string) => Promise<void>;
  get: (scope: Scope, id: string, opts?: QueryOpts) => Promise<StoreGetResp<T>>;
  query: (
    scope: Scope,
    queryString: string,
    mode: QueryMode,
    opts?: QueryOpts
  ) => Promise<EngineQueryResp<T>>;
  write: <T>(
    scope: Scope,
    node: IDNode<T>,
    opts?: DEngineStoreWriteOpts
  ) => Promise<void>;
}

export type DEngineMode = "exact" | "fuzzy";

export type DEngineOpts = {
  root: string;
  mode?: DEngineMode;
};

/**
 * Query: path based
 * Get: id based
 */
export interface DEngine {
  notes: NoteDict;
  schemas: SchemaDict;
  opts: Required<DEngineOpts>;

  /**
   * Load all nodes
   */
  init: () => Promise<void>;

  delete: (id: string) => Promise<void>;
  /**
   * Get node based on id
   * get(id: ...)
   */
  get: (
    scope: Scope,
    id: string,
    mode: QueryMode,
    opts?: QueryOpts
  ) => Promise<EngineGetResp<DNodeData>>;

  // getBatch: (scope: Scope, ids: string[]) => Promise<NodeGetBatchResp>;

  /**
   * Get node based on query
   * query(scope: {username: lukesernau}, queryString: "project", nodeType: note)
   * - []
   * - [Node(id: ..., title: project, children: [])]
   */
  query: (
    scope: Scope,
    queryString: string,
    mode: QueryMode,
    opts?: QueryOpts
  ) => Promise<EngineQueryResp<DNodeData>>;

  write: (
    scope: Scope,
    node: IDNode<DNodeData>,
    opts?: NodeWriteOpts
  ) => Promise<void>;
  // /**
  //  * Write node to db
  //  */
  // write: (scope: Scope, node: IDNode) => Promise<void>;

  // /**
  //  * Write list of nodes
  //  */
  // writeBatch: (scope: Scope, nodes: DNodeDict) => Promise<void>;
}
