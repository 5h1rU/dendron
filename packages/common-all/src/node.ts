/* eslint-disable no-loop-func */
import _ from "lodash";
import moment from "moment";
import minimatch  from "minimatch" ;
import YAML from "yamljs";
import { DendronError } from "./error";
import {
  DNodeData,
  DNodeDict,
  DNodeRawOpts,
  DNodeRawProps,
  IDNode,
  IDNodeOpts,
  IDNodeType,
  INote,
  INoteOpts,
  ISchema,
  ISchemaOpts,
  NoteData,
  NoteRawProps,
  SchemaData,

  SchemaDict, SchemaRawOpts,
  SchemaRawProps
} from "./types";
import { genUUID } from "./uuid";


const UNKNOWN_SCHEMA_ID = "_UNKNOWN_SCHEMA";

export class DNodeUtils {
  static basename(nodePath: string, extension?: boolean) {
    if (extension) {
      const idx = nodePath.lastIndexOf(".md");
      nodePath = nodePath.slice(0, idx);
    }
    return nodePath.split(".").slice(-1)[0];
  }

  static dirName(nodePath: string) {
    return nodePath
      .split(".")
      .slice(0, -1)
      .join(".");
  }

  static domainName(nodePath: string) {
    return nodePath.split("."[0]);
  }

  static findClosestParent(fpath: string, nodes: DNodeDict, opts?: {noStubs: boolean}): IDNode {
    const cleanOpts = _.defaults(opts, {noStubs: false});
    const dirname = DNodeUtils.dirName(fpath);
    if (dirname === "") {
      return nodes["root"];
    }
    const maybeNode = _.find(nodes, { fname: dirname });
    if (maybeNode && !(maybeNode?.stub && cleanOpts.noStubs)) {
      return maybeNode;
    } else {
      return DNodeUtils.findClosestParent(dirname, nodes);
    }
  }

  static isRoot(node: DNode): boolean {
    return node.id === "root";
  }

}

export type CreatePropsOpts = {
  returnExtra: boolean
}

export class DNodeRaw {
  static createProps<T>(nodeOpts: DNodeRawOpts<T>, opts?: CreatePropsOpts): DNodeRawProps<T> & {extra?: any} {
    const {
      id,
      desc,
      fname,
      updated,
      created,
      parent,
      stub,
      children,
      body,
      data,
      custom
    } = _.defaults(nodeOpts, {
      updated: moment.now(),
      created: moment.now(),
      id: genUUID(),
      desc: "",
      children: [],
      stub: false,
      parent: null,
      body: "",
      data: {},
      fname: null,
      custom: {}
    });
    const title = nodeOpts.title || fname;
    const nodeProps: DNodeRawProps<T> & {extra?: any} = {
      id,
      title,
      desc,
      fname,
      updated,
      created,
      parent,
      children,
      stub,
      body,
      data,
      custom,
    };
    if (opts?.returnExtra) {
      const extra = _.omit(nodeOpts, _.keys(nodeProps));
      nodeProps.extra = extra;
    }
    return nodeProps
  }
}

/*
Create Schema based on Minimal Props
- id: b111db5b-bc52-4977-893b-307522f89ea3
  title: "foo",
  parent: null
  children:
    - one
*/
export class SchemaNodeRaw {
  static createProps(opts: SchemaRawOpts): SchemaRawProps {
    opts.title = opts.title || opts.id;
    return DNodeRaw.createProps<SchemaData>(opts);
  }
}

type QuickPickItem = {
  label: string;

  /**
   * A human-readable string which is rendered less prominent in the same line. Supports rendering of
   * [theme icons](#ThemeIcon) via the `$(<name>)`-syntax.
   */
  description?: string;

  /**
   * A human-readable string which is rendered less prominent in a separate line. Supports rendering of
   * [theme icons](#ThemeIcon) via the `$(<name>)`-syntax.
   */
  detail?: string;

  /**
   * Optional flag indicating if this item is picked initially.
   * (Only honored when the picker allows multiple selections.)
   *
   * @see [QuickPickOptions.canPickMany](#QuickPickOptions.canPickMany)
   */
  picked?: boolean;

  /**
   * Always show this item.
   */
  alwaysShow?: boolean;
};

export abstract class DNode<T = DNodeData> implements IDNode<T>, QuickPickItem {
  public id: string;
  public title: string;
  public desc: string;
  public fname: string;
  public type: IDNodeType;
  public updated: string;
  public created: string;
  public parent: IDNode<T> | null;
  public children: IDNode<T>[];
  public body: string;
  public data: T;
  public label: string;
  public stub: boolean;
  public custom: any;

  constructor(opts: IDNodeOpts<T>) {
    const {
      id,
      title,
      desc,
      fname,
      type,
      updated,
      created,
      stub,
      body,
      data,
      children,
      custom
    } = _.defaults(
      opts,
      DNodeRaw.createProps(_.defaults(opts, { parent: null, children: [] }))
    );

    this.id = id;
    this.title = title || fname.split(".").slice(-1)[0];
    this.desc = desc;
    this.fname = fname;
    this.type = type;
    this.updated = updated;
    this.created = created;
    this.parent = opts.parent ? opts.parent : null;
    this.children = children;
    this.body = body;
    this.data = data;
    this.label = DNodeUtils.isRoot(this) ? "root" : this.logicalPath;
    this.stub = stub;
    this.custom = custom;
  }

  get domain(): DNode<T> {
    if (this.parent?.title === "root" || _.isNull(this.parent)) {
      return this;
    }
    return this.parent.domain;
  }

  get basename(): string {
    return DNodeUtils.basename(this.logicalPath);
  }

  get detail(): string {
    return "";
  }

  /**
   * Self and all children
   */
  get nodes(): DNode<T>[] {
    const out: DNode<T>[] = [this as DNode<T>].concat(
      this.children.map(c => c.nodes).flat()
    );
    return out;
  }

  /**
   * dot delimited path
   *  - for root node, its ""
   *  - for everything else, its the dot delimited name
   *  - used when showing query
   */
  get logicalPath(): string {
    if (this.fname === "root") {
      return "";
    } else {
      return this.fname;
    }
  }

  get path(): string {
    return this.fname;
  }

  abstract get url(): string;

  addChild(node: IDNode<T>) {
    // only add if new
    if (!this.children.some(ent => ent.id === node.id)) {
      this.children.push(node);
      node.parent = this;
    }
  }

  equal(node: IDNode<T>) {
    const props1 = this.toRawProps();
    const props2 = node.toRawProps();
    return _.every([
      _.isEqual(_.omit(props1, "body"), _.omit(props2, "body")),
      _.trim(props1.body) === _.trim(props2.body)
    ]);
  }

  // abstract match(identifier: any): boolean;

  renderBody(): string {
    return this.body;
  }

  toDocument() {
    return {
      document: {
        nodes: [
          {
            object: "block",
            type: "paragraph",
            nodes: [
              {
                object: "text",
                text: this.renderBody()
              }
            ]
          }
        ]
      }
    };
  }

  toRawProps(): DNodeRawProps<T> {
    const props = _.pick(this, [
      "id",
      "title",
      "desc",
      "type",
      "updated",
      "created",
      "body",
      "fname",
      "data",
      "stub",
      "custom",
    ]);
    let parent;
    if (this.parent?.title === "root") {
      parent = "root";
    } else {
      // TODO: this should never happen
      parent = this.parent?.id ?? "root";
    }
    const children = this.children.map(c => c.id);
    return { ...props, parent, children };
  }

  toRawPropsRecursive(): DNodeRawProps<T>[] {
    const parent: DNodeRawProps<T> = this.toRawProps();
    const children: DNodeRawProps<T>[] = this.children
      .map(
        (ch: DNode<T>) =>
          // @ts-ignore
          ch.toRawPropsRecursive()
        // eslint-disable-next-line function-paren-newline
      )
      .flat();
    // @ts-ignore
    const out = [parent].concat(children);
    return out.flat();
  }
  //othrow Error("to implement");
  //   return [];
  // }

  validate(): boolean {
    return true;
  }
}

export class Note extends DNode<NoteData> implements INote {
  public schemaId: string;
  public schema?: Schema;
  public schemaStub: boolean;

  static createStub(fname: string, opts?: Partial<INoteOpts>): Note {
    return new Note({ stub: true, fname, ...opts });
  }

  static createRoot(): Note {
    return new Note({ fname: "root", id: "root", title: "root" });
  }

  static fromSchema(dirpath: string, schema: Schema): Note {
    const fname = [dirpath, schema.title].join(".");
    const note = new Note({ fname, schemaStub: true, data: { schemaId: schema.id } });
    note.schema = schema;
    return note;
  }

  constructor(props: INoteOpts) {
    const cleanProps = _.defaults(props, {
      parent: null,
      children: [],
      schemaStub: false,
    });
    super({
      type: "note",
      ...cleanProps
    });
    this.schemaId = props?.data?.schemaId || "-1";
    this.schemaStub = cleanProps.schemaStub;
  }

  // vscode detail pane
  get detail(): string {
    const cleanPath = this.logicalPath
      .split(".")
      .slice(0, -1)
      .join(".");
    if (_.isEmpty(cleanPath)) {
      return this.logicalPath;
    } else {
      return cleanPath + ".";
    }
  }

  get description(): string | undefined {
    let schemaPrefix: string | undefined;
    let prefixParts = [];
    if (this.stub) {
      prefixParts.push("(stub)")
    }
    if (this.schema) {
      // case: unknown schema
      // eslint-disable-next-line no-use-before-define
      if (SchemaUtils.isUnkown(this.schema)) {
        prefixParts.push("$(squirrel) unknown");
        return prefixParts.join(" ");
      }

      // case: recognized schema
      prefixParts = ["$(repo)", this.schema.domain.title];
      // check if non-domain schema
      if (this.schema.domain.id !== this.schema.id) {
        prefixParts.push("$(breadcrumb-separator)");
        prefixParts.push(this.schema.title);
      }

      if (this.schemaStub) {
        prefixParts.push("(create from schema)");
      }
    }
    schemaPrefix = prefixParts.join(" ");
    return schemaPrefix;
  }

  get domain(): Note {
    return super.domain as Note;
  }

  get url(): string {
    return `/doc/${this.id}`;
  }
}

export class Schema extends DNode<SchemaData> implements ISchema {
  static createRoot() {
    const props = SchemaNodeRaw.createProps({ id: "root", fname: "root.schema" });
    return new Schema({ ...props, parent: null, children: [] });
  }

  static _UNKNOWN_SCHEMA: undefined|Schema = undefined;

  /**
   * This is attached to notes that are part of a domain with schema but
   * don't match any schema in it
   */
  static createUnkownSchema(): Schema {
    if (Schema._UNKNOWN_SCHEMA) {
      const props = SchemaNodeRaw.createProps({ id: UNKNOWN_SCHEMA_ID, fname: UNKNOWN_SCHEMA_ID, stub: true,
      created: "-1", updated: "-1" });
      Schema._UNKNOWN_SCHEMA = new Schema({ ...props, parent: null, children: [] });
    }
    return Schema._UNKNOWN_SCHEMA as Schema;
  }

  constructor(props: ISchemaOpts) {
    super({
      type: "schema",
      ..._.defaults(props, {
        parent: null,
        children: [],
        data: {}
      })
    });
  }

  get namespace(): boolean {
    return this.data?.namespace || false;
  }

  get logicalPath(): string {
    const part = this.namespace ? `${this.id}/*` : this.id;
    if (this.parent && this.parent.id !== "root") {
      const prefix = this.parent.logicalPath;
      return [prefix, part].join("/")
    } else {
      return part;
    }
  }

  get url(): string {
    return `/schema/${this.id}`;
  }

  match(note: Note): boolean {
    // TODO: simple version
    return this.title === note.basename;
  }

  renderBody() {
    const out = _.map(this.toRawPropsRecursive(), props => {
      return _.pick(props, [
        "id",
        "title",
        "desc",
        "children",
        "parent",
        "data",
        "fname"
      ]);
    });
    return ["```", YAML.stringify(out, undefined, 4), "```"].join("\n");
  }
}

const matchSchemaPropsToId = (
  id: string,
  props: SchemaRawProps[]
): SchemaRawProps => {
  const out = _.find(props, p => _.some([p.id === id]));
  if (_.isUndefined(out)) {
    throw Error(`no match found for ${id}, props: ${props}`);
  }
  return out;
};

// TODO:move to node
function getRoot(nodes: NoteRawProps[]) {
  // nodes: {nodes}
  const rootNode = _.find(
    nodes,
    ent => ent.title === "root" || _.isNull(ent.parent)
  );
  if (!rootNode) {
    throw new DendronError("no root node found");
  }
  const node = new Note({ ...rootNode, parent: null, children: [] });
  return { node, childrenIds: rootNode.children };
}

/**
 * From nodes, return a connected note tree
 */
export class NodeBuilder {
  getDomainsRoot<T extends DNodeData>(
    nodes: DNodeRawProps<T>[]
  ): DNodeRawProps<T>[] {
    // nodes: {nodes}
    const rootNodes = _.filter(nodes, ent => ent.parent === "root");
    if (_.isEmpty(rootNodes)) {
      throw new DendronError("no root node found");
    }
    return rootNodes;
  }

  toNote(item: NoteRawProps, parents: Note[], opts: { schemas: SchemaDict }) {
    const node = new Note({ ...item, parent: null, children: [] });
    if (node.schemaId) {
      node.schema = opts.schemas[node.schemaId];
    }
    const { parent: parentId, children } = item;
    const parent: Note = _.find(parents, { id: parentId }) as Note;
    // NOTE: parents don't get resolved until this is called
    parent.addChild(node);
    // eslint-disable-next-line no-use-before-define
    if (node.schemaId === UNKNOWN_SCHEMA_ID) {
      // TODO: FRAGILE, might not work with root stubs
      // const domainSchema = assertExists<Schema>(node.domain.children[0].schema as Schema, "note domain does not have schema");
      //const domainSchema = assertExists<Schema>(node.domain.schema as Schema, "note domain does not have schema");
      node.schema = Schema.createUnkownSchema();
    }
    return { node, parent, children };
  }

  toSchema(item: SchemaRawProps, parent: Schema, props: SchemaRawProps[]) {
    // DEBUG: item: {item}, parents: {parents}
    const node = new Schema({ ...item, parent, children: [] });
    item.children.forEach(chId => {
      const match = matchSchemaPropsToId(chId, props);
      return this.toSchema(match, node, props);
    });
    parent.addChild(node);
    return node;
  }

  buildNoteFromProps(props: NoteRawProps[], opts: { schemas: SchemaDict }): Note[] {
    const { node: rootNode, childrenIds } = getRoot(props);
    const out = [];
    out.push([rootNode]);

    let parentNodes = [rootNode];
    let nodeIds = childrenIds;

    while (!_.isEmpty(nodeIds)) {
      const currentNodes: Note[] = [];

      nodeIds = nodeIds
        .map((id: string) => {
          const nodeProps = props.find(ent => ent.id === id) as NoteRawProps;
          const { node, children } = this.toNote(nodeProps, parentNodes, opts);
          currentNodes.push(node);
          return children;
        })
        .flat();
      out.push(currentNodes);
      parentNodes = currentNodes;
    }
    return out.flat();
  }

  buildSchemaFromProps(props: SchemaRawProps[]) {
    const root = Schema.createRoot();
    const rootDomains: SchemaRawProps[] = this.getDomainsRoot<SchemaData>(
      props
    );
    let out = [root];
    rootDomains.forEach(rootRaw => {
      const domain = this.toSchema(rootRaw, root, props);
      out = out.concat(domain.nodes as Schema[]);
    });
    // DEBUG ctx: "parseSchema", out:
    return out;
  }
}

export class NoteUtils {
  /**
   * @param from
   * @param to
   */
  static createStubNotes(from: Note, to: Note): Note[] {
    const stubNodes: Note[] = [];
    const fromPath = from.logicalPath;
    const toPath = to.logicalPath;
    const index = toPath.indexOf(fromPath) + fromPath.length;
    const diffPath = _.trimStart(toPath.slice(index), ".").split(".");
    let stubPath = fromPath;
    let parent = from;
    // last element is node
    diffPath.slice(0, -1).forEach(part => {
      // handle starting from root, path = ""
      if (_.isEmpty(stubPath)) {
        stubPath = part;
      } else {
        stubPath += `.${part}`;
      }
      const n = Note.createStub(stubPath);
      stubNodes.push(n);
      parent.addChild(n);
      parent = n;
    });
    parent.addChild(to);
    return stubNodes;
  }
}

export class SchemaUtils {
  static isUnkown(schema: Schema) {
    return schema.id === UNKNOWN_SCHEMA_ID;
  }
  static matchNote(note: Note, schemas: SchemaDict): Schema{
    return _.find(_.values(schemas), schema => {
       return minimatch(note.path.replace(/\./g, '/'), schema.logicalPath);
     }) || Schema.createUnkownSchema();
  }
}
