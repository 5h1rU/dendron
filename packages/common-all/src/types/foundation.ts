import { URI } from "vscode-uri";
import { DVault } from "./workspace";

export type DLoc = {
  fname?: string;
  id?: string;
  vault?: DVault;
  uri?: URI;
  anchorHeader?: string;
};

export type DLink = {
  type: "ref" | "wiki" | "md" | "backlink";
  original: string;
  value: string;
  alias?: string;
  pos: {
    start: number;
    end: number;
  };
  from: DLoc;
  to?: DLoc;
};

export type DNodeType = "note" | "schema";
export type DNodePointer = string;

/**
 * Props are the official interface for a node
 */
export type DNodeProps<T = any, TCustom = any> = {
  /**
   * Unique id of a note
   */
  id: string;
  /**
   * Node title
   */
  title: string;
  /**
   * Node description
   */
  desc: string;
  /**
   * Node links (eg. backlinks, wikilinks, etc)
   */
  links: DLink[];
  /**
   * Name of the node. This corresponds to the name of the file minus the extension
   */
  fname: string;
  /**
   * Whether this node is a note or a schema
   */
  type: DNodeType;
  /**
   * Last updated
   */
  updated: number;
  /**
   * Created
   */
  created: number;
  /**
   * Determines whether this node is a {@link stub https://wiki.dendron.so/notes/c6fd6bc4-7f75-4cbb-8f34-f7b99bfe2d50.html#stubs}
   */
  stub?: boolean;
  /**
   @deprecated
   */
  schemaStub?: boolean;
  /**
   * Immediate parent
   */
  parent: DNodePointer | null;
  /**
   * Immediate children
   */
  children: DNodePointer[];
  data: T;
  /**
   * Body of the note
   */
  body: string;
  /**
   * Custom frontmatter
   */
  custom?: TCustom;
  schema?: { moduleId: string; schemaId: string };
  vault: DVault;
};

export type SchemaData = {
  namespace?: boolean;
  pattern?: string;
  template?: SchemaTemplate;
};

export type SchemaTemplate = {
  id: string;
  type: "snippet" | "note";
};

export type SchemaProps = DNodeProps<SchemaData>;
export type NoteProps = DNodeProps<any, any>;
