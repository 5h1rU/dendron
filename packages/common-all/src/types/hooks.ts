export type DHookEntry = {
  id: string;
  pattern: string;
  type: "js";
};

export type DHookDict = {
  onCreate: DHookEntry[];
};

export enum DHookType {
  onCreate = "onCreate",
}
