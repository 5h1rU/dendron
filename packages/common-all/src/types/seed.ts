export type SeedConfig = {
  name: string;
  publisher: string;
  license: string;
  root: string;
  description: string;
  contact?: SeedContact;
};

export type SeedContact = {
  name: string;
  email?: string;
  url?: string;
};

export enum SeedCommands {
  INFO = "info",
}
