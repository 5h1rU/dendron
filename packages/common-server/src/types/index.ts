import { NoteProps } from "@dendronhq/common-all";
import execa from "execa";
import _ from "lodash";

export type DHookFunction = (opts: {
  note: NoteProps;
  execa: typeof execa;
  _: typeof _;
}) => Promise<NoteProps>;
