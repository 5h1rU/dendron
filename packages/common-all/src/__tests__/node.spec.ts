import { DNodeRaw, NodeBuilder, Schema } from "../node";

import { ISchemaOpts } from "../types";

test("adds 1 + 2 to equal 3", () => {
  expect(1 + 2).toBe(3);
});

function setup() {
  const foo = new Schema({
    id: "foo.root",
    title: "foo.root",
    fname: "foo.root",
    parent: null
  });
  const fooChild = new Schema({
    id: "foo.child",
    title: "foo.child",
    fname: "foo.child",
    parent: null
  });
  const fooGrandChild = new Schema({
    id: "foo.grandchild",
    title: "foo.grandchild",
    fname: "foo.grandchild",
    parent: null
  });
  foo.addChild(fooChild);
  fooChild.addChild(fooGrandChild);
  return { foo, fooChild, fooGrandChild };
}

describe("schema", () => {
  let schemas: ReturnType<typeof setup>;

  beforeEach(() => {
    schemas = setup();
  });

  describe("domainRoot", () => {
    test("at root", () => {
      const { foo } = schemas;
      expect(foo.domain.id).toEqual(foo.id);
      expect(foo.renderBody()).toMatchSnapshot("foo_body");
    });

    test("at child", () => {
      const { fooChild, foo } = schemas;
      expect(fooChild.domain.id).toEqual(foo.id);
      expect(fooChild.renderBody()).toMatchSnapshot("fooChild_body");
    });

    test("at grand child", () => {
      const { fooGrandChild, foo } = schemas;
      expect(fooGrandChild.domain.id).toEqual(foo.id);
    });
  });
});

// describe("NodeBuilder", () => {
//   // let store: FileStorage;
//   let root: string;
//   beforeAll(() => {
//     root = setupTmpDendronDir();
//     // store = createFileStorage(root);
//   });

// test("convert from raw", async () => {
//   const resp = await store.query(createScope(), "**/*", {});
//   const rawProps = resp.data.map(n => n.toRawProps());
//   const p = new NodeBuilder();
//   const nodes = p.parse(rawProps);
//   const nodesRaw = _.sortBy(rawProps, [ent => ent.title]);
//   const nodesParsedRaw = _.sortBy(
//     nodes.map(n => n.toRawProps()),
//     [ent => ent.title]
//   );
//   expect(nodesParsedRaw).toEqual(nodesRaw);
// });
// });
