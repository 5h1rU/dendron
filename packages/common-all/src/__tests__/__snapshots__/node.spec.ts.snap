// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DNoteUtils findClosestParent, parent grandchild -> domain root: main 1`] = `
Object {
  "body": "",
  "children": Array [],
  "custom": Object {},
  "data": Object {},
  "desc": "",
  "fname": "baz",
  "parent": "root",
  "stub": false,
  "title": "baz",
  "type": "note",
}
`;

exports[`DNoteUtils findClosestParent, parent grandchild -> parent: main 1`] = `
Object {
  "body": "",
  "children": Array [
    "alpha",
  ],
  "custom": Object {},
  "data": Object {},
  "desc": "",
  "fname": "foo.one",
  "parent": "foo",
  "stub": false,
  "title": "one",
  "type": "note",
}
`;

exports[`DNoteUtils findClosestParent, parent grandchild -> root: main 1`] = `
Object {
  "body": "",
  "children": Array [
    "foo",
  ],
  "custom": Object {},
  "data": Object {},
  "desc": "",
  "fname": "root",
  "parent": "root",
  "stub": false,
  "title": "root",
  "type": "note",
}
`;

exports[`NoteUtils createStubNotes, root -> bar.one: barChild 1`] = `
Array [
  Object {
    "body": "",
    "children": Array [
      "alpha",
    ],
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "bar.one",
    "parent": "bar",
    "stub": false,
    "title": "one",
    "type": "note",
  },
  Object {
    "body": "",
    "children": Array [],
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "bar.one.alpha",
    "parent": "one",
    "stub": false,
    "title": "alpha",
    "type": "note",
  },
]
`;

exports[`Schema domainRoot at child: fooChild_body 1`] = `
"\`\`\`
-
    id: one
    title: yml
    desc: \\"\\"
    children: [alpha]
    parent: foo
    data: {}
    fname: foo.schema.yml
-
    id: alpha
    title: yml
    desc: \\"\\"
    children: []
    parent: one
    data: {}
    fname: foo.schema.yml

\`\`\`"
`;

exports[`Schema domainRoot at root: foo_body 1`] = `
"\`\`\`
-
    id: foo
    title: yml
    desc: \\"\\"
    children: [one]
    parent: root
    data: {}
    fname: foo.schema.yml
-
    id: one
    title: yml
    desc: \\"\\"
    children: [alpha]
    parent: foo
    data: {}
    fname: foo.schema.yml
-
    id: alpha
    title: yml
    desc: \\"\\"
    children: []
    parent: one
    data: {}
    fname: foo.schema.yml

\`\`\`"
`;

exports[`SchemaUtils note matchDomain: schemas 1`] = `
Array [
  Object {
    "body": "",
    "children": Array [
      "yml",
    ],
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo.schema.yml",
    "parent": "root",
    "stub": false,
    "title": "yml",
    "type": "schema",
  },
  Object {
    "body": "",
    "children": Array [
      "yml",
    ],
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo.schema.yml",
    "parent": "yml",
    "stub": false,
    "title": "yml",
    "type": "schema",
  },
  Object {
    "body": "",
    "children": Array [],
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo.schema.yml",
    "parent": "yml",
    "stub": false,
    "title": "yml",
    "type": "schema",
  },
  Object {
    "body": "",
    "children": Array [
      "yml",
    ],
    "custom": Object {},
    "data": Object {
      "namespace": true,
    },
    "desc": "",
    "fname": "bar.schema.yml",
    "parent": "root",
    "stub": false,
    "title": "yml",
    "type": "schema",
  },
  Object {
    "body": "",
    "children": Array [],
    "custom": Object {},
    "data": Object {
      "namespace": true,
    },
    "desc": "",
    "fname": "bar.schema.yml",
    "parent": "yml",
    "stub": false,
    "title": "yml",
    "type": "schema",
  },
]
`;

exports[`SchemaUtils string node: domain: schemas 1`] = `
Array [
  Object {
    "body": "",
    "children": Array [
      "yml",
    ],
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo.schema.yml",
    "parent": "root",
    "stub": false,
    "title": "yml",
    "type": "schema",
  },
  Object {
    "body": "",
    "children": Array [
      "yml",
    ],
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo.schema.yml",
    "parent": "yml",
    "stub": false,
    "title": "yml",
    "type": "schema",
  },
  Object {
    "body": "",
    "children": Array [],
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo.schema.yml",
    "parent": "yml",
    "stub": false,
    "title": "yml",
    "type": "schema",
  },
  Object {
    "body": "",
    "children": Array [
      "yml",
    ],
    "custom": Object {},
    "data": Object {
      "namespace": true,
    },
    "desc": "",
    "fname": "bar.schema.yml",
    "parent": "root",
    "stub": false,
    "title": "yml",
    "type": "schema",
  },
  Object {
    "body": "",
    "children": Array [],
    "custom": Object {},
    "data": Object {
      "namespace": true,
    },
    "desc": "",
    "fname": "bar.schema.yml",
    "parent": "yml",
    "stub": false,
    "title": "yml",
    "type": "schema",
  },
]
`;
